src/grain/grain.test.js

```diff
@@ -81,4 +77,36 @@ describe("src/grain/grain", () => {
       }
     });
   });
+
+  describe("multiplyFloat", () => {
+    it("behaves reasonably for tiny grain values", () => {
+      // $ExpectFlowError
+      expect(multiplyFloat(1n, 5)).toEqual(5n);
+    });
+    it("behaves reasonably for larger grain values", () => {
+      // $ExpectFlowError
+      expect(multiplyFloat(ONE, 2)).toEqual(2n * ONE);
+    });
+    it("has small error on large grain values", () => {
+      // To compare with arbitrary precision results, see:
+      // https://observablehq.com/@decentralion/grain-arithmetic
```

Comment:
I think this is an example that asks for a smoke test and a full test separation.
This being the inconclusive smoke test with a couple of fast-to-test examples.

The full test would fuzz this function with a substantial range of possible arguments and tests whether our specified margin of error holds true for all of these possibly unforeseen cases.

Reply:
I’m certainly open to something like that, with two caveats:

  - It’s not clear to me how to generate the expected values.
  - With the [exact implementation][1], we’re more worried about logic
	errors than precision errors.

On the first point: assuming that we don’t want to shell out to Python
or Java, I haven’t yet seen an “obviously correct” exact implementation
in JavaScript. Unfortunately, the results given by `decimal` in
@decentralion’s notebook are not actually correct. The `Decimal`
constructor, given a float, returns a decimal that is within an
f64-halfulp of the correct value but not within a decimal-halfulp of the
correct value. For instance,

```javascript
const decimal = require("decimal.js-light");
decimal(Math.PI).toPrecision(24, decimal.ROUND_HALF_DOWN)
```

is 3.14159265358979300000000, but the correct value is:

> 3.141592653589793115997963468544185161590576171875

The first three missing digits (“…115”) matter at attointeger scale. We
could resolve this concern if we could find a way to tell `decimal` to
give the correct answers.

On the second point: with the inexact `Number(grain)` implementation,
we’re worried about the worst-case error within an arbitrary interval of
floats—say, floats between `0.01` and `5.0`. This interval contains both
`1/1337` (pretty low error) and `Math.PI` (notably higher error), so a
sweep/smoke test over the interval would plausibly be quite valuable.

With the exact implementation, the edge cases to worry about aren’t
arbitrary float values within a given interval, but rather weird edge
cases of floats themselves: subnormals, zero mantissas, NaN payloads,
etc. The failure modes are more likely to have to do with the bit
manipulation than the algebra, because the algebra is uniform across all
floats. Thus, we’re worried about logic errors rather than precision
errors, and it’s not clear to me that fuzzing is an effective way to
address that.

[1]: https://github.com/sourcecred/sourcecred/commit/173be6b923d049895489a803bc353c47efdfff09#diff-14fdfc9807975908ca7d0b18bc2f3b76R174-R189

Reply:
> On the first point: assuming that we don’t want to shell out to Python
> or Java, I haven’t yet seen an “obviously correct” exact implementation
> in JavaScript.

Should we not have a good option available in JS, rather than shell out I would go for a static corpus, generating expected values for as strings in a JSON file for instance, parsing those as bigint.

> With the exact implementation, the edge cases to worry about aren’t
> arbitrary float values within a given interval, but rather weird edge
> cases of floats themselves: subnormals, zero mantissas, NaN payloads,
> etc.

That makes sense, my suggestion did not consider the exact implementation, just what was there in the PR commit so far. If we're going with lossy and want to be sure about our margin of error, we'll need more test coverage than 2 cases :] If we're ditching the lossy approach testing float edge cases makes more sense.